feature: Kobo Annotation to Hardcover.app sync

wolffshots opened the pull request:
This PR brings Kobo annotation syncing to CWA. If configured it changes the `readingservices_host` in the sync request to point to CWA (similar to what happens with `image_host`) and then requests to the Reading Services are intercepted, interpreted, added to  the relevant Hardcover Reading Journal (if the config allows) and the request is still proxied through to the actual Reading Services so that people with multiple Kobo's still get the same syncing functionality across them (I think).

---

## The features that are implemented are:
- syncing annotation from Kobo device to Hardcover.app - proxying requests still to retain normal functionality
  - handles creation of annotations (new `updatedAnnotations` on PATCH request body)
  - handles annotation deletions (in CWA db and the relevant journal entry on Hardcover - `deletedAnnotationIds` on PATCH request body)
  - handles syncing previous annotations on a book (`annotations` on the GET request response)
  - handles updating already existing entries if the note or highlighted text change on the above creation events
- opting in and out of Hardcover syncing 
    - globally - added annotation option
    - for specific books (opt out/blacklist) - added annotation and progress options
- split up the Dockerfile to allow for layer caching - I'm worried that this may cause issues if base layers never rebuild so may revert this once I am finished on the rest of these bits I'm working on but maybe will just look at it again soon
- add external IDs to books being synced
- only add books to Hardcover "want to read" if they don't exist on shelf add

---

I ended up reimplementing it and cleaning this PR and comments up

wolffshots commented:
I haven't added tests for the stuff I've added yet, was aiming to just get it functional for myself since asap since I just got a Kobo

wolffshots committed:
ci: speed up dockerfile building

wolffshots committed:
chore: add config for annotations

wolffshots committed:
feat: implement annotation and sync tracking

wolffshots committed:
feat: work out page progress from chapter position

wolffshots committed:
chore: implement blacklist for hardcover syncing

wolffshots committed:
feat: add intial processing for updating annotatations

wolffshots committed:
chore: prevent adding books already in user reads from shelf

wolffshots committed:
feat: process deleted annotations

wolffshots committed:
feat: process getting all annotations for a book

wolffshots committed:
chore: cleanup some trace statments

wolffshots committed:
chore: add missing proxy request

wolffshots committed:
feat: handle updating journal entries

wolffshots committed:
chore: defensively check hardcover details when mutating

wolffshots committed:
chore: proxy catch all route

Thovi98 commented:
This would be a nice feature to have! Thanks for the work!
I’ve just tried from this branch, but unfortunately annotations doesn’t appear on Hardcover (although reading progress is updated). I’ve enabled the Hardcover annotations in the settings and proxy to the Kobo store.

It seems that the logs only say: 
```
[2025-11-12 15:35:34,876]  INFO {cps.kobo:1106} Init
[2025-11-12 15:35:35,100]  INFO {cps.kobo:135} Kobo library sync request received
```
(My installation is a bare-metal installation without docker, so it can be my fault though).

wolffshots commented:
Heya @Thovi98
The fact that it is bare-metal rather than through docker shouldn't affect it
Could you please check if there are any entries from readingservices.py in your logs? If so what are they and if not could you enable DEBUG logging (`<your instance>/admin/config` > logfile config) and then try create or delete an annotation and then manually sync from the home page.
Worth also checking that the book you are expecting to sync has the relevant hardcover deets (`hardcover-edition`/`hardcover-id`/`hardcover-slug`)

Thovi98 commented:
Thanks @wolffshots !

Here are my debug logs:
```
[2025-11-12 18:14:39,320] DEBUG {cps.ub:72} Login and store session : 1a4933870f83e1f0c6a299e62070a3d9f456d27138e977b805ad51c731be7dd02f21b5e9a7d349d1a72d46cd9f8fac2d6aa307105ee941a8fe1007fb647bb957
[2025-11-12 18:14:48,232] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/c2fecbe8-5a84-4ac1-908e-03f5e780946b/prices (request is forwarded to kobo if configured)
[2025-11-12 18:14:50,416] DEBUG {cps.ub:72} Login and store session : 1a4933870f83e1f0c6a299e62070a3d9f456d27138e977b805ad51c731be7dd02f21b5e9a7d349d1a72d46cd9f8fac2d6aa307105ee941a8fe1007fb647bb957
[2025-11-12 18:14:52,705] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/affiliate (request is forwarded to kobo if configured)
[2025-11-12 18:14:52,756] DEBUG {cps.ub:72} Login and store session : 1a4933870f83e1f0c6a299e62070a3d9f456d27138e977b805ad51c731be7dd02f21b5e9a7d349d1a72d46cd9f8fac2d6aa307105ee941a8fe1007fb647bb957
[2025-11-12 18:14:52,757]  INFO {cps.kobo:1106} Init
[2025-11-12 18:14:52,789] DEBUG {cps.kobo:1024} Unimplemented User Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/user/profile (request is forwarded to kobo if configured)
[2025-11-12 18:14:52,855] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/books/subscriptions (request is forwarded to kobo if configured)
[2025-11-12 18:14:52,881] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/deals (request is forwarded to kobo if configured)
[2025-11-12 18:14:52,969] DEBUG {cps.ub:72} Login and store session : 1a4933870f83e1f0c6a299e62070a3d9f456d27138e977b805ad51c731be7dd02f21b5e9a7d349d1a72d46cd9f8fac2d6aa307105ee941a8fe1007fb647bb957
[2025-11-12 18:14:52,988]  INFO {cps.kobo:135} Kobo library sync request received
[2025-11-12 18:14:52,988] DEBUG {cps.kobo:136} SyncToken: 2025-11-04 11:32:40.964942,2025-11-12 12:48:29.298392,0001-01-01 00:00:00,2025-11-12 14:36:30.082431,2022-08-17 17:08:04.280904,eyJ0eXAiOjEsInZlciI6bnVsbCwicHR5cCI6IlN5bmNUb2tlbiJ9.eyJJbnRlcm5hbFN5bmNUb2tlbiI6ImV5SjBlWEFpT2pFc0luWmxjaUk2SW5ZeUlpd2ljSFI1Y0NJNklsTjVibU5VYjJ0bGJpSjkuZXlKVGRXSnpZM0pwY0hScGIyNUZiblJwZEd4bGJXVnVkSE1pT25zaVNYTkpibWwwYVdGc0lqcHVkV3hzTENKSFpXNWxjbUYwYVc5dVZHbHRaU0k2Ym5Wc2JDd2lWR2x0WlhOMFlXMXdJam9pTVRrd01DMHdNUzB3TVZRd01Eb3dNRG93TUZvaUxDSkRhR1ZqYTFOMWJTSTZiblZzYkN3aVNXUWlPaUl3TURBd01EQXdNQzB3TURBd0xUQXdNREF0TURBd01DMHdNREF3TURBd01EQXdNREFpZlN3aVJuVjBkWEpsVTNWaWMyTnlhWEIwYVc5dVJXNTBhWFJzWlcxbGJuUnpJanB1ZFd4c0xDSkZiblJwZEd4bGJXVnVkSE1pT25zaVNYTkpibWwwYVdGc0lqcHVkV3hzTENKSFpXNWxjbUYwYVc5dVZHbHRaU0k2Ym5Wc2JDd2lWR2x0WlhOMFlXMXdJam9pTWpBeU5TMHhNQzB4TVZReU1UbzBOam96TTFvaUxDSkRhR1ZqYTFOMWJTSTZJbnRjSWxOcmFYQkRjbVZoZEdWa1hDSTZabUZzYzJVc1hDSkRjbVZoZEdWa1EyOTFiblJjSWpvd0xGd2lRM0psWVhSbFpGd2lPbHRkTEZ3aVUydHBjRTF2WkdsbWFXVmtYQ0k2Wm1Gc2MyVXNYQ0pOYjJScFptbGxaRndpT2x0ZGZTSXNJa2xrSWpvaU1EQXdNREF3TURBdE1EQXdNQzB3TURBd0xUQXdNREF0TURBd01EQXdNREF3TURBd0luMHNJa1JsYkdWMFpXUkZiblJwZEd4bGJXVnVkSE1pT25zaVNYTkpibWwwYVdGc0lqcHVkV3hzTENKSFpXNWxjbUYwYVc5dVZHbHRaU0k2Ym5Wc2JDd2lWR2x0WlhOMFlXMXdJam9pTWpBeU5TMHhNQzB4TVZReU1UbzBOam96TTFvaUxDSkRhR1ZqYTFOMWJTSTZiblZzYkN3aVNXUWlPaUl3TURBd01EQXdNQzB3TURBd0xUQXdNREF0TURBd01DMHdNREF3TURBd01EQXdNREFpZlN3aVUybGtaV3h2WVdSbFpFVnVkR2wwYkdWdFpXNTBjeUk2ZXlKSmMwbHVhWFJwWVd3aU9tNTFiR3dzSWtkbGJtVnlZWFJwYjI1VWFXMWxJanB1ZFd4c0xDSlVhVzFsYzNSaGJYQWlPaUl5TURJMUxURXdMVEV4VkRJeE9qUTJPak16V2lJc0lrTm9aV05yVTNWdElqcHVkV3hzTENKSlpDSTZJakF3TURBd01EQXdMVEF3TURBdE1EQXdNQzB3TURBd0xUQXdNREF3TURBd01EQXdNQ0o5TENKRVpXeGxkR1ZrVTJsa1pXeHZZV1JsWkVWdWRHbDBiR1Z0Wlc1MGN5STZleUpKYzBsdWFYUnBZV3dpT201MWJHd3NJa2RsYm1WeVlYUnBiMjVVYVcxbElqcHVkV3hzTENKVWFXMWxjM1JoYlhBaU9pSXhPVEF3TFRBeExUQXhWREF3T2pBd09qQXdXaUlzSWtOb1pXTnJVM1Z0SWpwdWRXeHNMQ0pKWkNJNklqQXdNREF3TURBd0xUQXdNREF0TURBd01DMHdNREF3TFRBd01EQXdNREF3TURBd01DSjlMQ0pTWldGa2FXNW5VM1JoZEdWeklqcDdJa2x6U1c1cGRHbGhiQ0k2Ym5Wc2JDd2lSMlZ1WlhKaGRHbHZibFJwYldVaU9tNTFiR3dzSWxScGJXVnpkR0Z0Y0NJNklqSXdNalV0TVRBdE1URlVNakU2TkRZNk16TmFJaXdpUTJobFkydFRkVzBpT201MWJHd3NJa2xrSWpvaU1EQXdNREF3TURBdE1EQXdNQzB3TURBd0xUQXdNREF0TURBd01EQXdNREF3TURBd0luMHNJbFJoWjNNaU9uc2lTWE5KYm1sMGFXRnNJanB1ZFd4c0xDSkhaVzVsY21GMGFXOXVWR2x0WlNJNmJuVnNiQ3dpVkdsdFpYTjBZVzF3SWpvaU1qQXlOUzB4TUMweE1WUXlNVG8wTmpvek0xb2lMQ0pEYUdWamExTjFiU0k2Ym5Wc2JDd2lTV1FpT2lJd01EQXdNREF3TUMwd01EQXdMVEF3TURBdE1EQXdNQzB3TURBd01EQXdNREF3TURBaWZTd2lSR1ZzWlhSbFpGUmhaM01pT25zaVNYTkpibWwwYVdGc0lqcHVkV3hzTENKSFpXNWxjbUYwYVc5dVZHbHRaU0k2Ym5Wc2JDd2lWR2x0WlhOMFlXMXdJam9pTVRrd01DMHdNUzB3TVZRd01Eb3dNRG93TUZvaUxDSkRhR1ZqYTFOMWJTSTZiblZzYkN3aVNXUWlPaUl3TURBd01EQXdNQzB3TURBd0xUQXdNREF0TURBd01DMHdNREF3TURBd01EQXdNREFpZlN3aVVISnZaSFZqZEUxbGRHRmtZWFJoSWpwN0lrbHpTVzVwZEdsaGJDSTZiblZzYkN3aVIyVnVaWEpoZEdsdmJsUnBiV1VpT201MWJHd3NJbFJwYldWemRHRnRjQ0k2SWpJd01qVXRNVEV0TURSVU1qSTZNRFk2TlRWYUlpd2lRMmhsWTJ0VGRXMGlPbTUxYkd3c0lrbGtJam9pTTJJM05ERmtPRGN0T0RRek1DMDBZalprTFRrNFpUVXROV0prWTJZMU5tTTFOalJpSW4xOSIsIklzQ29udGludWF0aW9uVG9rZW4iOmZhbHNlfQ
[2025-11-12 18:14:52,988] DEBUG {cps.kobo:137} Download link format https://mydomain//kobo/1ac4688580df9551185ad1b6567695a1/download/%5Bbookid%5D/%5Bbookformat%5D
[2025-11-12 18:14:53,008] DEBUG {cps.kobo:201} Kobo Sync: books last modified: 2025-11-12 12:48:29.298392
[2025-11-12 18:14:53,013] DEBUG {cps.kobo:239} Kobo Sync: changed entries: 0
[2025-11-12 18:14:53,015] DEBUG {cps.kobo:243} Kobo Sync: selected to sync: 0
[2025-11-12 18:14:53,019] DEBUG {cps.kobo:291} Kobo Sync: remaining books to sync: 0
[2025-11-12 18:14:53,019] DEBUG {cps.kobo:295} Kobo Sync: rstate last modified: 2025-11-12 14:36:30.082431
[2025-11-12 18:14:53,020] DEBUG {cps.kobo:312} Kobo Sync: changed states: 1
[2025-11-12 18:14:53,119] DEBUG {cps.kobo:1024} Unimplemented User Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/user/wishlist (request is forwarded to kobo if configured)
[2025-11-12 18:14:53,502] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/books/series/270230b6-c27d-327f-951d-6d60c70fdc35 (request is forwarded to kobo if configured)
[2025-11-12 18:14:53,519] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/books/series/4b356a56-3b10-3c3f-bcdb-3fa4a4090aec (request is forwarded to kobo if configured)
[2025-11-12 18:14:53,581] DEBUG {cps.kobo:1024} Unimplemented User Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/user/recommendations (request is forwarded to kobo if configured)
[2025-11-12 18:14:54,118] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/c2fecbe8-5a84-4ac1-908e-03f5e780946b/nextread (request is forwarded to kobo if configured)
[2025-11-12 18:14:54,129] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/f94be5b2-b6bd-4337-8328-6b07a03d5b17/nextread (request is forwarded to kobo if configured)
[2025-11-12 18:14:54,140] DEBUG {cps.kobo:1066} Unimplemented Products Request received: https://mydomain/kobo/1ac4688580df9551185ad1b6567695a1/v1/products/f2ee1905-6b57-4170-9506-784f9749037a/nextread (request is forwarded to kobo if configured)
```
(if useful, the double slash (//) `Download link format https://mydomain` is not a typo)

I’ve `hardcover-edition` and `hardcover-id` with the same value (the edition of the book) and `hardcover-slug` with the actual book title

wolffshots commented:
Hmm if reading_services_host has been updated to your domain then the update to the sync code is updating that so I wonder why it's not seeing the new requests.

Resetting your Kobo probably isn't necessary but I can give you the instructions to get it into developer mode and capture the network request logs that it's doing and we'll see if maybe the requests are failing due to network conditions or something

crocodilestick committed:
feat: enhance Kobo sync functionality and improve code structure

crocodilestick committed:
Optimize Kobo annotation sync performance

- Implement `EpubProgressCalculator` for single-pass EPUB parsing to fix repeated I/O
- Move blacklist database check outside the annotation processing loop
- Refactor sync logic to reuse calculated progress and blacklist statusOptimize Kobo annotation sync performance

- Implement `EpubProgressCalculator` for single-pass EPUB parsing to fix repeated I/O
- Move blacklist database check outside the annotation processing loop
- Refactor sync logic to reuse calculated progress and blacklist status

crocodilestick committed:
Merge origin/main into pr/wolffshots/731

crocodilestick reviewed (commented):
@wolffshots thank you so much for this! Really amazing work, I just merged it into dev and it will be in the next release :)
